src/Codata/Musical/Colist/Infinite-merge.agda

```diff
@@ -116,107 +119,92 @@ private
-  from-injective (here (inj₁ _))  (here (inj₂ _))  ()
-  from-injective (here (inj₂ _))  (here (inj₁ _))  ()
-  from-injective (here (inj₂ p₁)) (here (inj₂ p₂)) eq     =
-    P.cong (here ∘ inj₂) $
-    drop-inj₁ $
-    Inverse.injective (Inv.sym (Any-⋎P _)) {x = inj₁ p₁} {y = inj₁ p₂} $
-    drop-there eq
-  from-injective (here (inj₁ _))  (there _)  ()
-  from-injective (here (inj₂ p₁)) (there p₂) eq
-    with Inverse.injective (Inv.sym (Any-⋎P _))
-                           {x = inj₁ p₁} {y = inj₂ (from p₂)}
-                           (drop-there eq)
-  ... | ()
-  from-injective (there _)  (here (inj₁ _))  ()
-  from-injective (there p₁) (here (inj₂ p₂)) eq
-    with Inverse.injective (Inv.sym (Any-⋎P _))
-                           {x = inj₂ (from p₁)} {y = inj₁ p₂}
-                           (drop-there eq)
-  ... | ()
-  from-injective (there {x = _ , xs} p₁) (there p₂) eq =
-    P.cong there $
-    from-injective p₁ p₂ $
-    drop-inj₂ $
-    Inverse.injective (Inv.sym (Any-⋎P xs))
-                      {x = inj₂ (from p₁)} {y = inj₂ (from p₂)} $
-    drop-there eq
-
-  -- The to function (defined as a right inverse of from).
-
-  Input = ∃ λ xss → Any P (merge xss)
-
-  Pred : Input → Set _
-  Pred (xss , p) = ∃ λ (q : Any Q xss) → from q ≡ p
-
-  to : ∀ xss p → Pred (xss , p)
-  to = λ xss p →
-    WF.All.wfRec (WF.Inverse-image.wellFounded size <′-wellFounded) _
-                 Pred step (xss , p)
+module _ {a p} {A : Set a} {P : A → Set p} where
+
+  Any-merge : ∀ xss → Any P (merge xss) ↔ Any (λ { (x , xs) → P x ⊎ Any P xs }) xss
+  Any-merge xss = inverse (proj₁ ∘ to xss) from (proj₂ ∘ to xss) to∘from
     where
-    size : Input → ℕ
-    size (_ , p) = index p
-
-    step : ∀ p → WF.WfRec (_<′_ on size) Pred p → Pred p
-    step ([]             , ())      rec
-    step ((x , xs) ∷ xss , here  p) rec = here (inj₁ p) , P.refl
-    step ((x , xs) ∷ xss , there p) rec
-      with Inverse.to (Any-⋎P xs) ⟨$⟩ p
-         | Inverse.left-inverse-of (Any-⋎P xs) p
-         | index-Any-⋎P xs p
-    step ((x , xs) ∷ xss , there .(Inverse.from (Any-⋎P xs) ⟨$⟩ inj₁ q)) rec | inj₁ q | P.refl | _   = here (inj₂ q) , P.refl
-    step ((x , xs) ∷ xss , there .(Inverse.from (Any-⋎P xs) ⟨$⟩ inj₂ q)) rec | inj₂ q | P.refl | q≤p =
-      Prod.map there
-               (P.cong (there ∘ _⟨$⟩_ (Inverse.from (Any-⋎P xs)) ∘ inj₂))
-               (rec (♭ xss , q) (s≤′s q≤p))
+    open P.≡-Reasoning
+
+    -- The from function.
+
+    Q = λ { (x , xs) → P x ⊎ Any P xs }
+
+    from : ∀ {xss} → Any Q xss → Any P (merge xss)
+    from (here (inj₁ p))        = here p
+    from (here (inj₂ p))        = there (Inverse.from (Any-⋎P _)  ⟨$⟩ inj₁ p)
+    from (there {x = _ , xs} p) = there (Inverse.from (Any-⋎P xs) ⟨$⟩ inj₂ (from p))
+
+    -- Some lemmas.
+
+    drop-there : ∀ {x xs} {p q : Any P _} →
+                 _≡_ {A = Any P (x ∷ xs)} (there p) (there q) → p ≡ q
+    drop-there P.refl = P.refl
+
```

Comment:
Is this `there-injective`? It could me moved to the toplevel.

Reply:
Yes, good idea. I originally thought it was `there` from `Data.Any` hence why I have those two new lemmas randomly.

Reply:
Weird, it already existed!

Reply:
Less work for you! Woohoo!